test/functional/p2p_erlay.py

```diff
@@ -0,0 +1,503 @@
+                self.handle_reconciliation_finalization(self.test_node, expected_success=False,
+                    expected_requested_txs=[])
+        else:
+            self.transmit_sketch(self.test_node, txs_to_sketch=mininode_txs, extension=False,
+                capacity=capacity, expected_announced_txs=[])
+            more_node_txs.extend(self.generate_txs(0, 1, 0, self.blocks, self)[1])
+            self.handle_extension_request(self.test_node)
+            more_node_txs.extend(self.generate_txs(0, 1, 0, self.blocks, self)[1])
+            if expected_success:
+                self.transmit_sketch(self.test_node, txs_to_sketch=mininode_txs, extension=True,
+                    capacity=capacity, expected_announced_txs=node_unique_txs)
+                self.handle_reconciliation_finalization(self.test_node, expected_success=True,
+                    expected_requested_txs=mininode_unique_txs)
+            else:
+                self.transmit_sketch(self.test_node, txs_to_sketch=mininode_txs, extension=True,
+                    capacity=capacity, expected_announced_txs=node_txs)
+                self.handle_reconciliation_finalization(self.test_node, expected_success=False,
+                    expected_requested_txs=[])
+
+        self.request_transactions(self.test_node, node_unique_txs)
+        # Check those additional transactions are not lost.
+        self.proceed_in_time(RECON_REQUEST_INTERVAL + 1)
+        self.receive_reqreconcil(self.test_node, expected_set_size=len(more_node_txs))
+        self.transmit_sketch(self.test_node, txs_to_sketch=[], extension=False, capacity=0, expected_announced_txs=more_node_txs)
+        self.handle_reconciliation_finalization(self.test_node, expected_success=False,
+            expected_requested_txs=[])
+
+    # The node being tested is a reconciliation initiator, the mininode is responder.
+    def test_recon_initiator(self):
+        MININODE_SALT = random.randrange(0xffffff)
+        self.blocks_offset = 0
+
+        # The first node is used to feed the node being tested with transactions.
+        self.test_node0 = self.nodes[0].add_p2p_connection(TestP2PConn(recon_version=1,
+            mininode_salt=MININODE_SALT, be_recon_requestor=True, be_recon_responder=False))
+        self.test_node0.wait_for_verack()
+
+        # These nodes will consume regular tx flood forwarding
+        for i in range(MAX_OUTBOUND_FLOOD_TO):
+            flooding_node = self.nodes[0].add_p2p_connection(TestP2PConn(
+                recon_version=0, mininode_salt=MININODE_SALT), node_outgoing=True)
+            flooding_node.wait_for_verack()
+
+        self.test_node = self.nodes[0].add_p2p_connection(TestP2PConn(recon_version=1,
+            mininode_salt=MININODE_SALT, be_recon_requestor=False, be_recon_responder=True), node_outgoing=True)
+        self.test_node.wait_for_verack()
+
+        self.blocks = self.nodes[0].generate(nblocks=1024)
```

Comment:
Why 1024? Is it an arbitrary number? I tested it using 500 blocks and it seems to work. Is there an explanation?

Reply:
Yeah, pretty arbitrary. I could generate them in `generate_transactions` function, but then it would force block sync across nodes, which is not ideal to test that transactions are not lost between scenarios (block sync could hide this effect by forcing tx sync)